Add llama3-llava-next-8b to llava_next conversion script

jamt9000 opened the pull request:
# What does this PR do?

Adds support for the [lmms-lab/llama3-llava-next-8b](https://huggingface.co/lmms-lab/llama3-llava-next-8b) model to the convert_llava_next_weights_to_hf.py script, along with an example prompt generated from the [llava_llama_3 conv_template](https://github.com/LLaVA-VL/LLaVA-NeXT/blob/6944062b9bb2e61c48436f1a65c3ea339095ec91/llava/conversation.py#L357) in the LLaVA-NeXT repo.

I've confirmed the model seems to convert without errors and the output seems similar (but not identical) to [the inference example from the LLaVA-NeXT repo](https://github.com/LLaVA-VL/LLaVA-NeXT/blob/inference/docs/LLaVA-NeXT.md#quick-start-with-huggingface), however I'm nor sure if the logic around the added tokens is correct.

The batched generation also seems to be off:
```
Batched generation...
Setting `pad_token_id` to `eos_token_id`:128009 for open-end generation.
['system\n\nYou are a helpful language and vision assistant. You are able to understand the visual content that the user provides, and assist the user with a variety of tasks using natural language.user\n\n\nWhat is shown in this image?assistant\n\n\nThe image appears to be a radar chart, also known as a spider chart or a web chart', '[INST] \nHow many cats are there? [/INST]\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n']
```

<!-- Remove if not applicable -->

Fixes https://github.com/huggingface/transformers/issues/31394


## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

@NielsRogge @zucchini-nlp

jamt9000 committed:
Add llama3-llava-next-8b to llava_next conversion script

Adds support for the lmms-lab/llama3-llava-next-8b model to the
convert_llava_next_weights_to_hf.py script, along with an example
prompt generated from the llava_llama_3 conv_template in the LLaVA-NeXT
repo.

zucchini-nlp reviewed (commented):
Hey! Thanks so much for adding support for new checkpoints ❤️  

Just a few comments:

- Can you also add the chat format [here in the docs](https://github.com/huggingface/transformers/blob/main/docs/source/en/model_doc/llava_next.md) to let users know which format to use. We don't have `chat_template` yet so it's better to indicate it in the docs for now
- Weird that batched generation didn't work. Can you verify that `tokenizer.padding_side="left"` or maybe it's the incorrect prompt format? 
- I am wondering if we need to add the other checkpoints, of 72b and 110b size. This is more of a question to @NielsRogge

NielsRogge commented:
Please also make sure to verify the outputs of the model against the original implementation, also on a logits level (ideally with `torch.assertallclose(original_logits[0,:3,:3], huggingface_logits[0,:3,:3])` on a dummy input). Getting sensible output is usually not enough

jamt9000 commented:
Currently doing a bunch of debugging to try to get near-identical logits. Findings so far are:

I'd added `<|begin_of_text|>` in the prompt example, but it looks like this gets added automatically in the HF processor, so the input_ids had two begin of text ids. I've also checked the 5-crop image preprocessing is completely identical.

After fixing that the logits are:

```py
# LLaVA-NeXT repo
tensor([[ -3.9648,   1.1396,   3.3145],
        [ -5.3398,  -1.5537,  -1.9512],
        [-12.3828, -10.6797,  -9.3047]], device='cuda:0',
       grad_fn=<SliceBackward0>)

# HF LlavaNextForConditionalGeneration
tensor([[ -3.9648,   1.1396,   3.3145],
        [ -5.3594,  -1.5654,  -1.9619],
        [-12.3750, -10.6797,  -9.3125]], device='cuda:2', dtype=torch.float32)
```

Which is close but not within the `atol=1e-4` in the script

jamt9000 committed:
Exclude <|begin_of_text|> from prompt example

This token gets added automatically, so it should not be included in the
prompt example.

jamt9000 commented:
Also making a note that LLama3 and Qwen actually have extra unused space in the embeddings, so we don't need to resize.

For LLama3 they are included in the tokenizer as `<|reserved_special_token_0|>` (you can see them all in https://huggingface.co/meta-llama/Meta-Llama-3-8B-Instruct/blob/main/tokenizer_config.json) and from the discussion here https://github.com/meta-llama/llama3/issues/77 it looks like they are available for use.

For Qwen, the unused space isn't already allocated in the tokenizer, but `len(tokenizer)==151646` whereas `config.text_config.vocab_size==152064` due to padding to a multiple of 128 https://github.com/QwenLM/Qwen/issues/553

jamt9000 committed:
Add llava-next-72b and llava-next-110b

Adds the Qwen-based LLaVA-Next models to the conversion script, along
with changes to load the models on multiple GPUs for inference.

whyiug commented:
hey, i wonder why this PR is in pending.

zucchini-nlp commented:
Sorry, totally forgot about this PR. I think the only thing left now is to get batched generation fixed, if not done yet, and create section in docs for new models, as outlined above

@jamt9000 let me know if you're stuck and need help or ping me if it's ready for review :)

jamt9000 committed:
Add llama3 and qwen prompt formats to docs

jamt9000 committed:
Chat prompt and padding side left for llama3 batched

jamt9000 commented:
I've added the prompt examples to the docs and checked that with the correct prompt and `padding_side = "left"` the batched generation is working (but with `padding_side = "right"` it still doesn't generate any answer to the cats question).

```
[
'system\n\nYou are a helpful language and vision assistant. You are able to understand the visual content that the user provides, and assist the user with a variety of tasks using natural language.user\n\n\nWhat is shown in this image?assistant\n\n\nThe image shows a radar chart, also known as a spider chart or a web chart, which',
'system\n\nYou are a helpful language and vision assistant. You are able to understand the visual content that the user provides, and assist the user with a variety of tasks using natural language.user\n\n\nHow many cats are there?assistant\n\n\nThere are two cats in the image.'
]
```

What I'm still stuck on is:

* I'm unsure if the llama3 logits need to be within a tighter tolerance to the reference
* Should the batched generation be working with padding=right? Why does it say "Setting `pad_token_id` to `eos_token_id`:128009 for open-end generation." when there's a padding token added?
* Whether it's necessary to resize the vocabulary (or if it could use a reserved token intead https://github.com/huggingface/transformers/pull/31395#issuecomment-2171694828)

zucchini-nlp commented:
@jamt9000 again sorry for delayed reply, was off on a vacation. Regarding the questions:

- Yes, `padding_side='left'` for batched inference is exactly what we need, as right padding is used only in training
- if we can make use of reserved tokens, then it's not necessary to resize vocab embeddings. Just make sure that tokenizer and model both have correct image token ids in their config. Simple test comparing with original implementation should be okay 
- Hmm, from the prev example seems like the higher logit difference is a result of using different dtypes in original and HF implementation. Afaik the original always uses bf16, so might interesting to check. In any case, the current state logits difference look okay to me

zucchini-nlp commented on docs/source/en/model_doc/llava_next.md:
````diff
@@ -66,6 +66,19 @@ The original code can be found [here](https://github.com/haotian-liu/LLaVA/tree/
 "<|im_start|>system\nAnswer the questions.<|im_end|><|im_start|>user\n<image>\nWhat is shown in this image?<|im_end|><|im_start|>assistant\n"
 ```
 
+llama3-llava-next-8b-hf requires the following format:
+
+```bash
+"<|start_header_id|>system<|end_header_id|>\n\nYou are a helpful language and vision assistant. You are able to understand the visual content that the user provides, and assist the user with a variety of tasks using natural language.<|eot_id|><|start_header_id|><|start_header_id|>user<|end_header_id|>\n\n<image>\nWhat is shown in this image?<|eot_id|><|start_header_id|>assistant<|end_header_id|>\ ...
+```
+
+llava-next-72b-hf and llava-next-110b-hf require the following format:
+
+```bash
+"<|im_start|>system\nYou are a helpful assistant.<|im_end|>\n<|im_start|>user\n<image>\nWhat is shown in this image?<|im_end|>\n<|im_start|>assistant\n"
+```
+
````
Great job! ❤️ 

We're trying to move to using `processor.apply_chat_template()`, so I will also add a `chat_template` for this models when uploading the weights

zucchini-nlp commented on src/transformers/models/llava_next/convert_llava_next_weights_to_hf.py:
```diff
@@ -151,28 +161,40 @@ def convert_llava_to_hf(model_id, pytorch_dump_folder_path, push_to_hub=False):
-    device = "cuda:2"
-    model.to(device)
+    if model_id in ["lmms-lab/llava-next-72b", "lmms-lab/llava-next-110b"]:
+        # For these big models need to do multi-gpu inference, so reload the model with device_map="auto" in order to use accelerate
+        # Is there a way to do this without saving and reloading the model?
+        model.save_pretrained("/tmp/llava_qwen")
+        model = LlavaNextForConditionalGeneration.from_pretrained("/tmp/llava_qwen", device_map="auto")
```
Hmm, imo it's okay to save and load the weights back. I would even say that `pytorch_dump_folder_path` doesn't have to be optional, so we save it there and then load back for inference

zucchini-nlp commented on src/transformers/models/llava_next/convert_llava_next_weights_to_hf.py:
```diff
@@ -279,9 +332,15 @@ def convert_llava_to_hf(model_id, pytorch_dump_folder_path, push_to_hub=False):
     url = "http://images.cocodataset.org/val2017/000000039769.jpg"
     cats_image = Image.open(requests.get(url, stream=True).raw)
 
+    if model_id == "lmms-lab/llama3-llava-next-8b":
+        cats_prompt = "<|start_header_id|>system<|end_header_id|>\n\nYou are a helpful language and vision assistant. You are able to understand the visual content that the user provides, and assist the user with a variety of tasks using natural language.<|eot_id|><|start_header_id|><|start_header_id|>user<|end_header_id|>\n\n<image>\nHow many cats are there?<|eot_id|><|start_header_id|>assistant ...
+        processor.tokenizer.padding_side = "left"
+    else:
+        cats_prompt = "[INST] <image>\nHow many cats are there? [/INST]"
+
     inputs = processor(
         images=[image, cats_image],
-        text=[prompt, "[INST] <image>\nHow many cats are there? [/INST]"],
+        text=[prompt, cats_prompt],
```
This can be chnaged to `[prompt, prompt]` so we don't have to manually match chat templates for the second prompt

zucchini-nlp reviewed (commented):
I left some comments. Let me know if you have bandwidth to address them, otherwise I can finish the PR for you and merge it next week :)

jamt9000 commented:
@zucchini-nlp Thanks! I'm on vacation this week, so do go ahead if you'd like to finish the PR yourself!

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_31395). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

zucchini-nlp committed:
update

zucchini-nlp committed:
Merge branch 'main' into add-llava-llama3-8b-to-conversion-script

zucchini-nlp commented:
Ready for review, made final changes and uploaded the weights on the Hub, along with their chat templates

amyeroberts commented on src/transformers/models/llava_next/convert_llava_next_weights_to_hf.py:
```diff
@@ -305,16 +364,27 @@ def convert_llava_to_hf(model_id, pytorch_dump_folder_path, push_to_hub=False):
-        Path(pytorch_dump_folder_path).mkdir(exist_ok=True)
-        model.save_pretrained(pytorch_dump_folder_path)
-        processor.save_pretrained(pytorch_dump_folder_path)
+    import os
+
+    from huggingface_hub import HfApi
```
imports should be at the top of the file

zucchini-nlp replied:
same, removed

amyeroberts commented on src/transformers/models/llava_next/convert_llava_next_weights_to_hf.py:
```diff
@@ -305,16 +364,27 @@ def convert_llava_to_hf(model_id, pytorch_dump_folder_path, push_to_hub=False):
+
+    from huggingface_hub import HfApi
+
+    repo_id = model_id.split("/")[-1]
+    print(f"Pushing to repo llava-hf/{repo_id}-hf")
+
+    api = HfApi()
+    for file in os.listdir(pytorch_dump_folder_path):
+        api.upload_file(
+            path_or_fileobj=f"{pytorch_dump_folder_path}/{file}",
+            path_in_repo=file,
+            repo_id=f"llava-hf/{repo_id}-hf",
+            repo_type="model",
+        )
```
We should only optionally push to the hub

amyeroberts commented on src/transformers/models/llava_next/convert_llava_next_weights_to_hf.py:
```diff
@@ -305,16 +364,27 @@ def convert_llava_to_hf(model_id, pytorch_dump_folder_path, push_to_hub=False):
+
+    from huggingface_hub import HfApi
+
+    repo_id = model_id.split("/")[-1]
```
Let's use a better name here - this isn't really the "repo_id", it's the checkpoint name

amyeroberts commented on src/transformers/models/llava_next/convert_llava_next_weights_to_hf.py:
```diff
@@ -151,28 +162,47 @@ def convert_llava_to_hf(model_id, pytorch_dump_folder_path, push_to_hub=False):
+    del state_dict
+    gc.collect()
+
+    # Load everything back for inference tests in float23 because prev script was written as that
```
```suggestion
    # Load everything back for inference tests in float32 because prev script was written as that
```

amyeroberts commented on src/transformers/models/llava_next/convert_llava_next_weights_to_hf.py:
```diff
@@ -151,28 +162,47 @@ def convert_llava_to_hf(model_id, pytorch_dump_folder_path, push_to_hub=False):
+    # Load everything back for inference tests in float23 because prev script was written as that
+    # Though it's mostly loaded in fp16 as original weights are in fp16
+    model = LlavaNextForConditionalGeneration.from_pretrained(
+        pytorch_dump_folder_path, torch_dtype=torch.float32, device_map="auto"
```
I don't think you need to specify `torch_dtype` here is you're loading in float32 - this will always happen automatically unless `torch_dtype="auto"` or a dtype which isn't float32 is specified

amyeroberts reviewed (commented):
Thanks for adding! 

Some general comments e.g. not pushing to the hub automatically

zucchini-nlp committed:
Update src/transformers/models/llava_next/convert_llava_next_weights_to_hf.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

zucchini-nlp commented on src/transformers/models/llava_next/convert_llava_next_weights_to_hf.py:
```diff
@@ -305,16 +364,27 @@ def convert_llava_to_hf(model_id, pytorch_dump_folder_path, push_to_hub=False):
-        repo_id = model_id.split("/")[-1]
-        model.push_to_hub(f"llava-hf/{repo_id}-hf")
-        processor.push_to_hub(f"llava-hf/{repo_id}-hf")
+    # if push_to_hub:
+    #     repo_id = model_id.split("/")[-1]
+    #     print(f"Pushing to repo llava-hf/{repo_id}-hf")
+    #     model.push_to_hub(f"llava-hf/{repo_id}-hf")
+    #     processor.push_to_hub(f"llava-hf/{repo_id}-hf")
```
Yes, my bad. Pushed changes that were used only locally. Will revert to pushing with `push_to_hub` and not `HfApi()`

zucchini-nlp commented on src/transformers/models/llava_next/convert_llava_next_weights_to_hf.py:
```diff
@@ -328,11 +398,14 @@ def convert_llava_to_hf(model_id, pytorch_dump_folder_path, push_to_hub=False):
     )
     parser.add_argument(
-        "--pytorch_dump_folder_path", default=None, type=str, help="Path to the output PyTorch model directory."
+        "--pytorch_dump_folder_path", type=str, required=True, help="Path to the output PyTorch model directory."
```
We want to save the model and then load again, so that it's loaded by automatically inferring how to split weights on each device/cpu/disk depending on how much memory available.

Especially required in big models that we're adding, they need multi-gpu inference but we can't init model from config in multi-gpu setting

zucchini-nlp committed:
Update src/transformers/models/llava_next/convert_llava_next_weights_to_hf.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

zucchini-nlp committed:
remove code

zucchini-nlp committed:
better naming

zucchini-nlp commented:
@amyeroberts ready for review, addresses the comments

amyeroberts reviewed (approved):
Thanks for adding and iterating!

zucchini-nlp commented:
Thanks @jamt9000 💛  The converted model weights are already uploaded to llava-hf org in the hub!
